### PR TITLE
Improves ActiveRecord::Base.connection_handler.while_preventing_writes documentation [ci skip]

### LINKED ISSUE #39132: ActiveRecord::Base.connection_handler.while_preventing_writes does not prevent all writes
The `ActiveRecord::Base.connection_handler.while_preventing_writes` method does not prevent all writes as the name suggests.

This should be mentioned in the method documentation, particularly because in the [original Changelog](https://github.com/eileencodes/rails/blob/f39d72d5267baed1000932831cda98503d1e1047/activerecord/CHANGELOG.md) the author @eileencodes does mention this.

### Description

While looking for a way to safely expose a sql query parameter I stumbled upon the `while_preventing_writes` method. I was surprised that I wouldn't need a another readonly DB user. Not quite sure if this would really work, I did some digging and found that, at least in Postgresql, this method is not sufficient to block all writes. 

### Steps to reproduce

All of the following queries do write to a Postgresql database in a `while_preventing_writes` block:

#### Explain analyze

Postgresql let's you not only explain a query, you can also [explain and analyze](https://www.postgresql.org/docs/current/sql-explain.html) a query which actually **runs** it and also analyzes it:

```ruby
ActiveRecord::Base.connection_handler.while_preventing_writes do
  ActiveRecord::Base.connection.execute "EXPLAIN ANALYZE INSERT INTO roles_users (role_id, user_id) VALUES (1, 1234)"
  ActiveRecord::Base.connection.execute "EXPLAIN ANALYZE UPDATE users SET role = 'admin' WHERE id = 10"
  ActiveRecord::Base.connection.execute "EXPLAIN ANALYZE DELETE FROM users WHERE id = 26"
end
```

#### Nested comments

Postgres has a [peculiar comment syntax](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-COMMENTS) which lets you nest comments within other comments:

```
/* comment with nesting: /* nested comment */ */
```

That is why these comments look like SELECT queries to the regular expression [READ_QUERY](https://github.com/rails/rails/blob/7cd59448ba7b841b47cf8fdbdab55053e56ba787/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb#L23) which is supposed to guard against writes, but to the SQL parser they are comments:

```ruby
ActiveRecord::Base.connection_handler.while_preventing_writes do
  ActiveRecord::Base.connection.execute "/*/**/SELECT*/ INSERT INTO roles_users (role_id, user_id) VALUES (1, 1234)"
  ActiveRecord::Base.connection.execute "/*/**/SELECT*/ UPDATE users SET role = 'admin' WHERE id = 10"
  ActiveRecord::Base.connection.execute "/*/**/SELECT*/ DELETE FROM users WHERE id = 26"
end
```

#### Select into

Finally postgres let's you create tables using the [`SELECT INTO` statement](https://www.postgresql.org/docs/12/sql-selectinto.html), which starts the query with a `SELECT` but then goes on to write the result into a new table. While it cannot write into an existing table, technically it's still a write query:

```ruby
ActiveRecord::Base.connection_handler.while_preventing_writes do
  ActiveRecord::Base.connection.execute "SELECT * INTO users_new FROM users"
end
```

### Expected behavior

The writing queries are not applied to the database.

### Actual behavior

The queries are written to the database.  The [regular](https://github.com/rails/rails/blob/7cd59448ba7b841b47cf8fdbdab55053e56ba787/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb#L23) [expression](https://github.com/rails/rails/blob/7cd59448ba7b841b47cf8fdbdab55053e56ba787/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L69) which is supposed to guard against write queries does not catch the write statements.

### Discussion

I suspect that parsing SQL with regexes is like parsing HTML with regexes: it cannot work. I suspect that @eileencodes was aware of the limitations of this regular expression approach and mentioned them in the Changelog at the time.

I did spend a fair amount of time trying to fix these issues and I believe that a somewhat more complex Regex will be able to catch some of these issues, but not all.

Considering that the nested comments can have any level of depth, I don't think that they can be solved with Regexes. You will need to parse the SQL and strip out the comments entirely to safeguard against those.

The `SELECT INTO` issue is complex as well, as there can be a lot of statements between the `SELECT` and the `INTO` some of which may be strings or comments, so the regular expression approach is probably not going to work here as well. 

I have tests and a fix for the `EXPLAIN ANALYZE` queries, if you want me to I can open a pull request for those. I will add a pull request for the documentation change.

### System configuration

* Rails master
* ruby 2.7.0p0
* vagrant@rails-dev-box

### PR DESCRIPTION
**Pull request by eikes**
Address #39132 "ActiveRecord::Base.connection_handler.while_preventing_writes does not prevent all writes"

Not all write queries which are wrapped in a `while_preventing_writes` block are
blocked. For example in Postgres these are valid queries which are actually
executed:

    ActiveRecord::Base.connection_handler.while_preventing_writes do
      ActiveRecord::Base.connection.execute "EXPLAIN ANALYZE DELETE FROM users WHERE id = 26"
      ActiveRecord::Base.connection.execute "/*/**/SELECT*/ UPDATE users SET role = 'admin' WHERE id = 10"
      ActiveRecord::Base.connection.execute "SELECT * INTO users_new FROM users"
    end

The documentation should mention, that this method is not supposed to catch all
writes, but rather the usual writes as they would be created by ActiveRecord.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 1]
```diff
@@ -1018,11 +1018,16 @@ def prevent_writes=(prevent_writes) # :nodoc:
...
       #
       # In some cases you may want to prevent writes to the database
       # even if you are on a database that can write. `while_preventing_writes`
-      # will prevent writes to the database for the duration of the block.
+      # will prevent most writes to the database for the duration of the block.
+      #
+      # Note: `while_preventing_writes` does not necessarily catch all writes,
+      # it's meant as a safeguard against accidental writes. Writes as created
+      # by ActiveRecord will most likely be caught. Handcrafted write queries
+      # may still be possible though.
```
Comment (eileencodes) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:1030]: This statement makes it sound like there isn't an explicit list of the kinds of queries that writes are prevented for. Can you write this more clearly and reference the list of disallowed queries?

**Review by eileencodes — CHANGES_REQUESTED** [review]

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb (hunk 1 above)
Comment (eikes) [reply activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:1030]: Well there really isn't any write query you can't do in Postgresql, you can make them all run as I pointed out in the issue #39133. Or to say it differently, there aren't any disallowed queries when you prepend your query with a `/*/**/SELECT*/` comment.

Comment (eileencodes) [reply activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:1030]: If you want true replicas you should use a readonly user on your replica. Rails can only do so much guess work for you.

The new documentation is just not clear enough to be something we can merge. All the dbs have an explicit list, even if there are ways around that. The new docs make it sound like Rails just blocks writes when it feels like it, but it's really more about how you craft your queries.

Comment (eikes) [reply activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:1030]: OK, maybe we can add something along the lines of: 

> Note: `while_preventing_writes` is not as safe as a separate readonly database user and it does not make any guarantees against SQL injections, it is meant to provide a safeguard against accidental writes by the Rails application itself.

**Comment by eikes** [comment]
I find it somewhat frustrating, that I put considerable effort in the issue into pointing out this potentially dangerous misnomer and it does not get addressed. The change is a minuscule addition to the documentation which may save someone from relying on this method to prevent SQL injections.

**Comment by eileencodes** [comment]
I have been _incredibly_ busy. I'm sorry that I haven't had time to further review this, but at the moment, it's not quite where I think it's mergeable. Will try to get to it this week.

**Comment by eileencodes** [comment]
Rather than have you do another round of editing, I've pushed a commit that clarifies the documentation. I added you as co-author so you get the credit. Sorry for the delay in reviewing.

Here's the commit 82217c21703d7b94a61c778748e9ccd354fd105c

**Comment by eileencodes** [comment]
I also backported this change to 6-0-stable in bba75d106cc49dae266c8b833a4156754f7e737b

**Pull request closed without merging**